[Infrastructure UI] Show processes on a single hosts

jennypavlova opened the pull request:
Closes #150907

## Summary
This PR adds the Processes tab to the single host flyout. The component is already implemented in Inventory so we can reuse it here 

## Testing
- Go to hosts view
- Open the flyout for any host to see the single host details
- Click on the processes tab 
⚠️ If you want to see the processes summary on top (where the total processes are displayed) you need inside your metricbeat modules yml configuration to include the `process_summary` so your config should include:
```
metricbeat.modules:
  - module: system
    metricsets:
      ..... other data  .......
      - process # Per process metrics
      - process_summary # Process summary
    ..... other data  .......
```
<img width="1913" alt="image" src="https://user-images.githubusercontent.com/14139027/228534978-c38437e4-4279-4ad4-9fc8-5222cbd15c2e.png">

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

jennypavlova commented:
> I was playing around and noticed that the tabs reload completely when you switch between them, is this intended?

@mohamedhamed-ahmed Thanks for checking it! 🙂  It works the same in inventory - you load the data when you select the tab. Also, we want to reload it if the time range is changed so maybe if we want to persist the data we should also compare the dates to make sure that the data matches the correct requested time. 
We can add new issues in the future and optimize the loading but I am currently using the same component as in inventory so it works the same way.

jennypavlova commented:
> Can you actually do this with the flyout opened?
Then the flyout loading time will increase if we load the whole content at once or that wasn't the idea 🤔 For now as we use the same components for the processes the whole loading logic is added there - so when we change something it's also affecting inventory. But we can think about some refactoring steps there to keep the last result while the flyout is still open also for inventory (in case of the same time range) probably in another issue so it's clear what we are doing 🙂 .

> also if you have some filter queries in the process tabs they get deleted when navigating which might also be another thing to file an issue for.

That's a good point! Also, we discussed in [another PR](https://github.com/elastic/kibana/pull/152956#issuecomment-1478077851) that we should persist the opened tab in the URL and then we can load the last tab the user visited - so it's a good idea to persist the process filters as well when we do that so if the user reopens the flyout the filters will be still there.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/components/node_details/tabs/processes/index.tsx:
```diff
@@ -29,7 +31,11 @@ import { SummaryTable } from './summary_table';
+  node: InfraWaffleMapNode | HostNodeRow;
+}
+
+const TabComponent = ({ currentTime, node, nodeType }: ProcessesProps) => {
```
In the future, we might want to start moving components that are now used in both Hosts View and Inventory UI to `public/components/inventory` as they are now common components. Ideally, we shouldn't create interdependencies between pages.

jennypavlova replied:
That's a really good suggestion, thanks! I would add a separate issue for that as it already has some imports from metrics explorer and infra inside so it will be a bigger refactoring.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/host_details_flyout/flyout.tsx:
```diff
@@ -9,15 +9,18 @@ import React, { useMemo, useState } from 'react';
 }
 
-const flyoutTabs = [MetadataTab];
+const flyoutTabs = [MetadataTab, ProcessesTab('hostsView-flyout-tabs-processes')];
```
As said in another suggestion, here we could create a new `ProcessesTab` with its own `data-subj-test` or use a spread operator. Also, maybe `MetadataTab` could be camelCased? It's actually an object.

crespocarlos commented on x-pack/test/functional/apps/infra/hosts_view.ts:
```diff
@@ -150,6 +154,67 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
+          expect(processValue).to.eql(313);
+        });
+
+        it('should render processes table', async () => {
```
Would it be interesting to validate the content of a row? Just to make sure it's rendering some content?

jennypavlova replied:
I added a check for the table body and the expand button click (if we don't have the data the table body won't be present - there is a message instead)

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/components/node_details/overlay.tsx:
```diff
@@ -47,7 +47,14 @@ export const NodeContextPopover = ({
+  const tabConfigs = [
+    MetricsTab,
+    LogsTab,
+    ProcessesTab('infraNodeTabsProcesses'),
```
Same as another comment. Or maybe use a spread operator and keep `ProcessesTab` with the bare minimum necessary to render a tab?

```
{ ...ProcessesTab, 'data-subj-test': infraNodeTabsProcesses }
```

Since the other tab representations are not functions but objects - I don't know why their names are capitalized either, as they are not components or classes - having this `ProcessTab` as a function seems a little bit confusing to me.

tonyghiani commented on x-pack/test/functional/page_objects/infra_hosts_view.ts:
```diff
@@ -28,6 +28,10 @@ export function InfraHostsViewProvider({ getService }: FtrProviderContext) {
       return await testSubjects.click('euiFlyoutCloseButton');
     },
 
+    async clickProcessesFlyoutTab() {
+      return await testSubjects.click('hostsView-flyout-tabs-processes');
+    },
```
**nit**: the `return await` is redundant, return directly the promise:
```js
clickProcessesFlyoutTab() {
  return testSubjects.click('hostsView-flyout-tabs-processes');
},
```

jennypavlova replied:
Ah, sorry about all of those (I was copying/pasting stuff around from different places and forgot to remove the wait after 🤦‍♀️ )

tonyghiani commented on x-pack/test/functional/apps/infra/hosts_view.ts:
```diff
@@ -150,6 +154,67 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
+          const processesTotalValue =
+            await pageObjects.infraHostsView.getProcessesTabContentTotalValue();
+          const processValue = await processesTotalValue.getVisibleText();
+          expect(processValue).to.eql(313);
```
**nit**: avoid coercion, if the visible text is a string go for
```js
expect(processValue).to.equal('313');
```

tonyghiani commented on x-pack/test/functional/apps/infra/constants.ts:
```diff
@@ -24,6 +24,8 @@ export const DATES = {
       withoutData: '10/09/2018 10:00:00 PM',
       min: '2018-10-17T19:42:21.208Z',
       max: '2018-10-17T19:58:03.952Z',
+      processesDataStartDate: '2023-03-28T18:20:00.000Z',
+      processesDataEndDate: '2023-03-28T18:21:00.000Z',
```
**question:** why don't we use a single date range for all the dataset? Having multiple time range can complicate our tests and fake data inconsistencies. We could keep the min/max timerange and adapt the esArchiver data to be in that range.

jennypavlova replied:
This is a bit of a special case I think - I want to see **only** the data coming from this data set so I used different dates - the background is that if we have other hosts from another dataset they are incomplete and don't include the `process.pid` so once the process tab is opened it will return an error (maybe some sort of filtering can be applied to avoid that I was just thinking that having only the host with the correct data for this test returned will prevent flaky results ) - so this way I know that only the host from this dataset will be returned. 
Another option was to edit all the existing data/mappings in the existing dataset but that will be a lot of manual editing and risk breaking existing tests.

tonyghiani commented on x-pack/test/functional/apps/infra/hosts_view.ts:
```diff
@@ -150,6 +154,67 @@ export default ({ getPageObjects, getService }: FtrProviderContext) => {
 
+    describe('#Single host Flyout', () => {
+      before(async () => {
+        await Promise.all([
```
**nit**: in this case `Promise.all` is not necessary since we want to run sequentially this setup actions.
Also, `Promise.all` expect an array of promises, so using `await` inside it to resolve them is not what is meant for.

jennypavlova commented on x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_flyout_tab_id.ts:
```diff
@@ -0,0 +1,40 @@
+import { constant, identity } from 'fp-ts/lib/function';
+import { useUrlState } from '../../../../utils/use_url_state';
+
+const FLYOUT_TAB_ID_URL_STATE_KEY = 'flyoutTabId';
```
Added in this [commit](https://github.com/elastic/kibana/pull/153866/commits/41824589188e1cf48ea4ff4964aa302004cc7ab5)

jennypavlova replied:
I update the url state implementation and also persisted the search text in this [commit](https://github.com/elastic/kibana/pull/153866/commits/c51f57754a4a94206faaea13def5050a9cd7da6a)

tonyghiani commented on x-pack/plugins/infra/public/pages/metrics/inventory_view/components/node_details/tabs/processes/summary_table.tsx:
```diff
@@ -44,8 +44,8 @@ export const SummaryTable = ({ processSummary, isLoading }: Props) => {
-          <EuiFlexItem>
-            <EuiDescriptionList compressed={true}>
+          <EuiFlexItem key={field}>
+            <EuiDescriptionList data-test-subj="infra_processes_summary_table_item" compressed>
```
The convention applies to the attribute, independently of what use we make of it so that it has the same structure everywhere in the app.

jennypavlova replied:
Thank you for explaining that, I switched to a camel case test subject.

jennypavlova commented:
@crespocarlos @mohamedhamed-ahmed @tonyghiani Thank you for the review ❤️ Most of the requested changes are applied and I change a bit the structure/functionality so to summarize:
- I added the URL persisting logic we discussed in the other [PR](https://github.com/elastic/kibana/pull/152956#issuecomment-1478077851) So I decided to do it here instead of creating a new issue for it as it would improve the UX here
- I also added a separate processes component to help persist the search text/filter inside the processes tab - most of the processes logic still need to be extracted to be in a shared folder as mentioned [here](https://github.com/elastic/kibana/pull/153866#discussion_r1152887095) and I added an [issue](https://github.com/elastic/kibana/issues/154164) for that
- Left the inventory TabComponent inside processes unchanged 
- Refactored and updated the tests

https://user-images.githubusercontent.com/14139027/229145536-eb492e18-b55d-4a34-9ecf-4da2af5c0b92.mov

Please let me know in case I miss something or if you have any suggestions 🙂

tonyghiani commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/hosts_table.tsx:
```diff
@@ -22,11 +22,11 @@ export const HostsTable = () => {
   });
 
-  const clickedItem = items.find(({ uuid }) => uuid === clickedItemUuid);
+  const clickedItem = items.find(({ id }) => id === clickedItemId);
```
**nit**: just realized now that also the current item (e.g. `selectedHost`) is derived from internals of the `useHostTable` hook, you can so lift this computation into the hook and memoize the value of the selected host (say we have 100 hosts, we don't want to iterate all of them on each render if the selection doesn't change), so that the hook directly return the selected host or undefined if not selected

jennypavlova replied:
Good idea, I moved that to the hook and memoized it 👍

tonyghiani commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/host_details_flyout/processes/processes.tsx:
```diff
@@ -0,0 +1,179 @@
+
+export const Processes = ({ currentTime, node, nodeType }: ProcessesProps) => {
+  const [hostFlyoutOpen, setHostFlyoutOpen] = useHostFlyoutOpen();
+  const [searchBarState, setSearchBarState] = useState<Query>(
```
**nit:** use the lazy init for this state to prevent parsing the `hostFlyoutOpen.searchFilter` unnecessarily after the first render.

jennypavlova replied:
Thanks for the suggestion! done 👍

tonyghiani commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/host_details_flyout/processes/processes.tsx:
```diff
@@ -0,0 +1,179 @@
+              name: 'State',
+              operator: 'exact',
+              multiSelect: false,
+              options: Object.entries(STATE_NAMES).map(([value, view]: [string, string]) => ({
+                value,
+                view,
+              })),
```
**nit**: this is a static value that has no dependencies from the component lifecycle, you can map it at a module level once and pass it here to avoid recreating this on each render.

jennypavlova replied:
yes, you are right, I moved that outside 👍

tonyghiani commented:
Hey @jennypavlova, good job, I just left a couple more suggestions and the we should be good to go 🚀
